include/util/alias.hpp

```diff
@@ -0,0 +1,112 @@
+/*
+
+Copyright (c) 2016, Project OSRM contributors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+Redistributions of source code must retain the above copyright notice, this list
+of conditions and the following disclaimer.
+Redistributions in binary form must reproduce the above copyright notice, this
+list of conditions and the following disclaimer in the documentation and/or
+other materials provided with the distribution.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef ALIAS_HPP
+#define ALIAS_HPP
+
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+namespace osrm
+{
+
+template <typename From, typename Tag> struct Alias;
+template <typename From, typename Tag>
+inline std::ostream &operator<<(std::ostream &stream, const Alias<From, Tag> &inst);
+
+template <typename From, typename Tag> struct Alias final
```

Comment:
If you make `Alias` a bit smarter and special case for primitive types vs. non-primitive types, you can inherit from non-primitive types and :sparkles: we have proper newtype semantics e.g. for vectors:

https://github.com/mapbox/node-or-tools/blob/master/src/types.h#L13-L23

compiler will complain if you mix things up, conversion is still possible to its base type